keps/sig-node/3983-drop-in-configuration/README.md

```diff
@@ -0,0 +1,315 @@
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Add support for a drop-in configuration directory for the Kubelet. This directory will be in `/etc/kubernetes/kubelet.conf.d` and configuration files will be processed in alphanumeric order. Establishment of conventions for configuration processing will be done, and further work can be done to add this support for other components.
+
+
+## Motivation
+
+A common pattern for software configuration in linux is support for a drop-in configuration directory. The location of this directory is often based on a corresponding configuration file. For instance, `/etc/security/limits` can be overridden by files in `/etc/security/limits.d`. This pattern is useful for a number of reasons, though a large motivation here is to allow files to be owned by a single owner. If multiple processes are vying for changing the same file, then they could stamp over each other's changes and possibly race against each other, creating TOCTOU problems.
+
+Components in Kubernetes can similarly be configured by multiple entities and preventing races between them is cumbersome. There has been past work in the Kubelet to have a Dynamic Configuration, but resolving between multiple entities and a last known good state was also complicated. Since the Kubelet is the node agent, and is often distributed as a package on the host operating system along with the container runtime, configuring it similarly to other host processes seems clear. This paves the path for continuing the pattern of drop-in configuration for the Kubelet.
+
+
+### Goals
+
+* Add support for a `/etc/kubernetes/kubelet.conf.d` directory, which will override the configuration for the Kubelet located at `/etc/kubernetes/kubelet.conf`
+* Extend API machinery code to reconcile multiple configuration files.
+* Define Kubernetes best-practices for configuration definitions, similarly to [API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md).
+
+### Non-Goals
+
+* Add support for drop-in configuration for Kubernetes components other than the Kubelet.
+
+## Proposal
+
+This proposal aims to add support for a drop-in configuration directory for the kubelet in `/etc/kubernetes/kubelet.conf.d`. Users are able to specify individually configurable kubelet config snippets in files, formatted in the same way as the existing kubelet.conf file. The kubelet will process the configuration provided in `/etc/kubernetes/kubelet.conf.d` in alphanumeric order:
+
+
+1. If no other configuration for the subfield(s) exist, append to the base configuration
+2. If the subfield(s) exists in the base configuration at `/etc/kubernetes/kubelet.conf` file or another file in `/etc/kubernetes/kubelet.conf.d` with lesser alphanumeric ordering, overwrite it
```

Comment:
Are you planning to add a `--config-dir` or similar flag to specify the directory? Also, AFAIU there currently is no default kubelet config file path like `/etc/kubernetes/kubelet.conf`, are you planning to specify a default for that (changing the current default behavior)?

Reply:
I would say the behavior should mirror that of the configuration file, which looks like there's no flag for it, but there is a default

Reply:
oop, I seem to have missed the glaringly obvious `--config` option, so yeah there should be a `--config-dir` option I think

Reply:
There is `--config` for specifying the kubelet config file but that's empty by default IIRC

Reply:
Thanks for the discussion! So as far as I understand, there's:

1. the kubelet.conf file generated with default configuration: https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/kubelet-integration/
2. the additional customizations you can make with a kubeletconfig file, passed via the flag: https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/

So in the base case with no additional specifications (which this enhancement wouldn't change) you would end up with the configurations in `/etc/kubernetes/kubelet.conf`

I can update the enhancement to allow specification of a configuration directory. One question I would have is, if you were to specify both config-dir and config flags, should those have a priority order?

Let me know if my assumptions are wrong anywhere, thanks!